# Lab 2. Search
# Task 2.3 Maze Solver
## 1.  Problem Descriptions
1.  **State:** Starting position  at (x, y) position labelled "O".
2.  **Action:** In orde to meet the boundary conditions the label "O" can move up, down, left and right.
3.  **Goal Test:** Check if label "O" is at exit labeled"X".
4.  **Path Cost** Equal step cost
## Implementation and Results

In [ ]:
!pip install simpleai
from simpleai.search import SearchProblem, astar, greedy, breadth_first, depth_first, uniform_cost
import math

# **Group 1 - cost value 1**(per move)

In [ ]:
MAP = """
##############################
#         #              #   #
# ####    ########       #   #
#  o #    #              #   #
#    ###     ####   ######   #
#         ####      #        #
#            #  #   #   #### #
#     ######    #       # x  #
#        #      #            #
##############################
"""
MAP = [list(x) for x in MAP.split("\n") if x]

COSTS = {
    "up": 1.0,
    "down": 1.0,
    "left": 1.0,
    "right": 1.0,
    # "up left": 1.4,
    # "up right": 1.4,
    # "down left": 1.4,
    # "down right": 1.4,
}


In [ ]:

class Maze(SearchProblem):

    def __init__(self, board):
        self.board = board
        self.goal = (0, 0)
        for y in range(len(self.board)):
            for x in range(len(self.board[y])):
                if self.board[y][x].lower() == "o":
                    self.initial = (x, y)
                elif self.board[y][x].lower() == "x":
                    self.goal = (x, y)

        super(Maze, self).__init__(initial_state=self.initial)

    def actions(self, state):
        actions = []
        for action in list(COSTS.keys()):
            newx, newy = self.result(state, action)
            if self.board[newy][newx] != "#":
                actions.append(action)
        return actions

    def result(self, state, action):
        x, y = state

        if action.count("up"):
            y -= 1
        if action.count("down"):
            y += 1
        if action.count("left"):
            x -= 1
        if action.count("right"):
            x += 1

        new_state = (x, y)
        return new_state

    def is_goal(self, state):
        return state == self.goal

    def cost(self, state, action, state2):
        return COSTS[action]

    def heuristic(self, state):
        x, y = state
        gx, gy = self.goal
        return math.sqrt((x - gx) ** 2 + (y - gy) ** 2)

Astar search method is used to search the graph from initial state label "O" to label "X" on graph

In [ ]:
problem = Maze(MAP)
result = astar(problem, graph_search=True)
path = [x[1] for x in result.path()]
for y in range(len(MAP)):
    for x in range(len(MAP[y])):
        print('.' if (x,y) in path[1:-1] else MAP[y][x], end='')
    print()
print("A* Search: moves=%s, cost=%s." %(result.depth, result.cost))

##############################
#         #              #   #
# ####    ########       #   #
#  o.#    # ........     #   #
#   .### ....####  .######   #
#   ......####     .#        #
#            #  #  .#   #### #
#     ######    #  .....#.x  #
#        #      #      ...   #
##############################
A* Search: moves=33, cost=33.0.


Greedy Best-First search method is used to search the graph from initial state label "O" to label "X" on graph.

In [ ]:
problem = Maze(MAP)
result = greedy(problem, graph_search=True)
path = [x[1] for x in result.path()]
for y in range(len(MAP)):
    for x in range(len(MAP[y])):
        print('.' if (x,y) in path[1:-1] else MAP[y][x], end='')
    print()
print("Greedy Best-First* Search: moves=%s, cost=%s." %(result.depth, result.cost))

##############################
#         #              #   #
# ####    ########       #   #
#  o.#    #              #   #
#   .###     ####   ######   #
#   ......#### .....#        #
#        ....# .#  .#   #### #
#     ######....#  .....#.x  #
#        #      #      ...   #
##############################
Greedy Best-First* Search: moves=33, cost=33.0.


Breadth-First search method is used to search the graph from initial state label "O" to label "X" on graph.

In [ ]:
problem = Maze(MAP)

result = breadth_first(problem, graph_search=True)
path = [x[1] for x in result.path()]
for y in range(len(MAP)):
    for x in range(len(MAP[y])):
        print('.' if (x,y) in path[1:-1] else MAP[y][x], end='')
    print()
print("Breadth-First Search: moves=%s, cost=%s." %(result.depth, result.cost))

##############################
#         #              #   #
# ####    ########       #   #
#  o #    #              #   #
#  . ###     ####   ######   #
#  .      ####....  #        #
#  ..........#. #.  #   #### #
#     ######... #.      #.x  #
#        #      #.........   #
##############################
Breadth-First Search: moves=33, cost=33.0.


Depth-First search method is used to search the graph from initial state label "O" to label "X" on graph.

In [ ]:
problem = Maze(MAP)
result = depth_first(problem, graph_search=True)
path = [x[1] for x in result.path()]
for y in range(len(MAP)):
    for x in range(len(MAP[y])):
        print('.' if (x,y) in path[1:-1] else MAP[y][x], end='')
    print()
print("Depth-First Search: moves=%s, cost=%s." %(result.depth, result.cost))

##############################
#         #              #   #
# ####    ########       #   #
#  o.#    #              #   #
#   .###     ####   ######   #
#   ......#### .....#        #
#        ....# .#  .#   #### #
#     ######....#  .....# x  #
#        #      #      ....  #
##############################
Depth-First Search: moves=33, cost=33.0.


Uniform-Cost search method is used to search the graph from initial state label "O" to label "X" on graph>


In [ ]:
problem = Maze(MAP)
result = uniform_cost(problem, graph_search=True)
path = [x[1] for x in result.path()]
for y in range(len(MAP)):
    for x in range(len(MAP[y])):
        print('.' if (x,y) in path[1:-1] else MAP[y][x], end='')
    print()
print("Uniform-Cost Search: moves=%s, cost=%s." %(result.depth, result.cost))

##############################
#         #              #   #
# ####    ########       #   #
#  o #    #              #   #
#  . ###     ####   ######   #
#  .....  ####..... #        #
#      ......#. # . #   #### #
#     ######... # ..    #.x  #
#        #      #  .......   #
##############################
Uniform-Cost Search: moves=33, cost=33.0.


# **Group 2 - cost value 1.4**(per move)


In [ ]:
MAP = """
##############################
#         #              #   #
# ####    ########       #   #
#  o #    #              #   #
#    ###     ####   ######   #
#         ####      #        #
#            #  #   #   #### #
#     ######    #       # x  #
#        #      #            #
##############################
"""
MAP = [list(x) for x in MAP.split("\n") if x]

COSTS = {    
    "up": 1.0,
    "down": 1.0,
    "left": 1.0,
    "right": 1.0,
    "up left": 1.4,
    "up right": 1.4,
    "down left": 1.4,
    "down right": 1.4,
}


In [ ]:
class Maze(SearchProblem):

    def __init__(self, board):
        self.board = board
        self.goal = (0, 0)
        for y in range(len(self.board)):
            for x in range(len(self.board[y])):
                if self.board[y][x].lower() == "o":
                    self.initial = (x, y)
                elif self.board[y][x].lower() == "x":
                    self.goal = (x, y)

        super(Maze, self).__init__(initial_state=self.initial)

    def actions(self, state):
        actions = []
        for action in list(COSTS.keys()):
            newx, newy = self.result(state, action)
            if self.board[newy][newx] != "#":
                actions.append(action)
        return actions

    def result(self, state, action):
        x, y = state

        if action.count("up"):
            y -= 1
        if action.count("down"):
            y += 1
        if action.count("left"):
            x -= 1
        if action.count("right"):
            x += 1

        new_state = (x, y)
        return new_state

    def is_goal(self, state):
        return state == self.goal

    def cost(self, state, action, state2):
        return COSTS[action]

    def heuristic(self, state):
        x, y = state
        gx, gy = self.goal
        return math.sqrt((x - gx) ** 2 + (y - gy) ** 2)


Astar search method is used to search the graph from initial state label "O" to label "X" on graph

In [ ]:
problem = Maze(MAP)
result = astar(problem, graph_search=True)
path = [x[1] for x in result.path()]
for y in range(len(MAP)):
    for x in range(len(MAP[y])):
        print('.' if (x,y) in path[1:-1] else MAP[y][x], end='')
    print()
print("A* Search: moves=%s, cost=%s." %(result.depth, result.cost))

##############################
#         #              #   #
# ####    ########       #   #
#  o #    #  ....        #   #
#   .###  ...####.  ######   #
#    .....####    . #        #
#            #  #  .#   #### #
#     ######    #   ....#.x  #
#        #      #       .    #
##############################
A* Search: moves=23, cost=26.999999999999993.


Greedy Best-First search method is used to search the graph from initial state label "O" to label "X" on graph.

In [ ]:
problem = Maze(MAP)
result = greedy(problem, graph_search=True)
path = [x[1] for x in result.path()]
for y in range(len(MAP)):
    for x in range(len(MAP[y])):
        print('.' if (x,y) in path[1:-1] else MAP[y][x], end='')
    print()
print("Greedy Best-First* Search: moves=%s, cost=%s." %(result.depth, result.cost))

##############################
#         #              #   #
# ####    ########       #   #
#  o #    #              #   #
#   .###     ####   ######   #
#    .    ####  .   #        #
#     ...... # .#.  #   #### #
#     ######... # ......#.x  #
#        #      #       .    #
##############################
Greedy Best-First* Search: moves=23, cost=26.999999999999996.


Breadth-First search method is used to search the graph from initial state label "O" to label "X" on graph.

In [ ]:
problem = Maze(MAP)

result = breadth_first(problem, graph_search=True)
path = [x[1] for x in result.path()]
for y in range(len(MAP)):
    for x in range(len(MAP[y])):
        print('.' if (x,y) in path[1:-1] else MAP[y][x], end='')
    print()
print("Breadth-First Search: moves=%s, cost=%s." %(result.depth, result.cost))

##############################
#         #              #   #
# ####    ########       #   #
#  o #    #  ....        #   #
#   .###  ...####.  ######   #
#    .....####    . #        #
#            #  #  .#   #### #
#     ######    #   ....# x  #
#        #      #       ..   #
##############################
Breadth-First Search: moves=23, cost=26.999999999999993.


Depth-First search method is used to search the graph from initial state label "O" to label "X" on graph.

In [ ]:
problem = Maze(MAP)
result = depth_first(problem, graph_search=True)
path = [x[1] for x in result.path()]
for y in range(len(MAP)):
    for x in range(len(MAP[y])):
        print('.' if (x,y) in path[1:-1] else MAP[y][x], end='')
    print()
print("Depth-First Search: moves=%s, cost=%s." %(result.depth, result.cost))

##############################
# ....    #              #   #
#.####.   ########       #   #
#. o # .  #              #   #
#.  .###.    ####   ######   #
# .  .   .####  .   #        #
#.    .   .. # .#.  #   #### #
#. . .######. . # . . . #.x  #
# . .    #   .  #  . . ..    #
##############################
Depth-First Search: moves=38, cost=49.59999999999997.


Uniform-Cost search method is used to search the graph from initial state label "O" to label "X" on graph.

In [ ]:
problem = Maze(MAP)
result = uniform_cost(problem, graph_search=True)
path = [x[1] for x in result.path()]
for y in range(len(MAP)):
    for x in range(len(MAP[y])):
        print('.' if (x,y) in path[1:-1] else MAP[y][x], end='')
    print()
print("Uniform-Cost Search: moves=%s, cost=%s." %(result.depth, result.cost))

##############################
#         #              #   #
# ####    ########       #   #
#  o #    #  ....        #   #
#   .###  ...####.  ######   #
#    .....####    . #        #
#            #  #  .#   #### #
#     ######    #   ....# x  #
#        #      #       ..   #
##############################
Uniform-Cost Search: moves=23, cost=26.999999999999993.


## Discussions

In this task, we have used 5 search algorithms.  The comparison shows that all of these algortihm can be used to determien shortest path.  Both moves and cost were the same across all serach methods.  

Also, we see that while having cost per move set at varying cost "value," there were no significant observable final cost difference between both groups. 

### Astar results comparison

Ananlysis of Heuristics behavior Summary:
  Group 2 test results
*  Cost per move = 1.4
*  Moves = 23
*  Cost = 26.99
Ananlysis of Heuristics behavior Summary:
  Group 1 test results
*  Cost per move = 1.0
*  Moves = 33
*  Cost = 33 

The cost for group 1 is much higher than group 2 which suggest that the estimated cost is greater than the cost to move.  Group 2 expanded from 4 to 8 moves whcih gave astar a more efficient way.  Hence, the estimated cost may be lower or equal to the moving(actual) cost in group 2 which acconts for the shorter distance.  

### Greedy Best-First

Ananlysis of Heuristics behavior Summary:
  Group 1 test results
*  Cost per move = 1.0
*  Moves = 33
*  Cost = 33 
Ananlysis of Heuristics behavior Summary:
  Group 2 test results
*  Cost per move = 1.4
*  Moves = 23
*  Cost = 26.99
We see that for both Gready Best-first and A* search for group 2 are the same.  This is only possible where the estimated cost is highly relative and in this case equal to actual cost.  In a sense A * search has turn into a Gready best-first search.   

### Depth-First search
Ananlysis of Heuristics behavior Summary:
  Group 1 test results
*  Cost per move = 1.0
*  Moves = 33
*  Cost = 33 
Ananlysis of Heuristics behavior Summary:
  Group 2 test results
*  Cost per move = 1.4
*  Moves = 38
*  Cost = 49.60

Given more direction, this search methode visited more nodes compared to group 1.  
### Uniform-cost
Ananlysis of Heuristics behavior Summary:
  Group 1 test results
*  Cost per move = 1.0
*  Moves = 33
*  Cost = 33 
Ananlysis of Heuristics behavior Summary:
  Group 2 test results
*  Cost per move = 1.4
*  Moves = 23
*  Cost = 26.99

Uniform cost all share the same results for both group 1 and group 2.  Since Uniform-cost uses the cheapest cost to get to the destination node, by having an additional 4 moves in group 2 improved its efficiency in cost andnodes visited.  Uniform

In conclusion, Uniform-cost, greedy best-first and A*  all share the same results.  hence, an accurate estimate that is 100% correct will ensure we gett he shortest path quickly.  
